File: classifyByAS.pl
use strict;
use warnings;
 
# check number of arguments passed is correct 
my $num_args = $#ARGV + 1; 
if ($num_args != 2) { 
    print "\nUsage: classifyByAS.pl Project SWID\n"; 
    exit; 
} 
 
# assign arguments to script variables 
# example input: 6816873 
my $mouseBam="/.mounts/labs/gsiprojects/gsi/Xenograft-Classifier/data/seqware_prov_report/projects/${ARGV[0]}/bwa/mm10/${ARGV[1]}.bam"; 
my $humanBam="/.mounts/labs/gsiprojects/gsi/Xenograft-Classifier/data/seqware_prov_report/projects/${ARGV[0]}/bwa/hg19/${ARGV[1]}.bam"; 
# my $outputFile="/.mounts/labs/gsiprojects/gsi/Xenograft-Classifier/data/seqware_prov_report/bams/${ARGV[0]}/hg19/notMouse/twoReads/${ARGV[1]}.sam"; # change to bam if possible 

# output files
(open my $GRAFT, ">", "/.mounts/labs/gsiprojects/gsi/Xenograft-Classifier/data/seqware_prov_report/projects/${ARGV[0]}/bwa/hg19/classifyByAS_results/${ARGV[1]}/graft.txt");
(open my $HOST, ">", "/.mounts/labs/gsiprojects/gsi/Xenograft-Classifier/data/seqware_prov_report/projects/${ARGV[0]}/bwa/hg19/classifyByAS_results/${ARGV[1]}/host.txt");
(open my $BOTH, ">", "/.mounts/labs/gsiprojects/gsi/Xenograft-Classifier/data/seqware_prov_report/projects/${ARGV[0]}/bwa/hg19/classifyByAS_results/${ARGV[1]}/both.txt");
(open my $NEITHER, ">", "/.mounts/labs/gsiprojects/gsi/Xenograft-Classifier/data/seqware_prov_report/projects/${ARGV[0]}/bwa/hg19/classifyByAS_results/${ARGV[1]}/neither.txt");
(open my $AMBIGUOUS, ">", "/.mounts/labs/gsiprojects/gsi/Xenograft-Classifier/data/seqware_prov_report/projects/${ARGV[0]}/bwa/hg19/classifyByAS_results/${ARGV[1]}/ambiguous.txt");
 
# open bams
(open my $MOUSE,"samtools view -F 256  $mouseBam |") || die "unable to open mouse bam: $mouseBam";  
(open my $HUMAN,"samtools view -F 256 $humanBam |") || die "unable to open human bam: $humanBam"; 
 
# (open my $HUMANHEADER,"samtools view -H $humanBam |") || die "unable to open human header"; 
# my @header=<$HUMANHEADER>; 
# close $HUMANHEADER; 
# (open my $OUT,">","$outputFile") || die "unable to open $outputFile"; 
# print $OUT join("",@header); 
 
# counters
my $totalCount=0; 
my $hostCount=0; 
my $graftCount=0; 
my $bothCount=0; 
my $neitherCount=0;
my $ambiguousCount=0; 

# read variables
my $mouserec1;
my $mouserec2;
my $humanrec1;
my $humanrec2;

# mapping type
my $mmapped;
my $hmapped;

while($mouserec1=<$MOUSE>){

        $mouserec2=<$MOUSE>;
        chomp $mouserec1;
        chomp $mouserec2;
        my @mfields1=split /\t/,$mouserec1;
        my @mfields2=split /\t/,$mouserec2;

        my $humanrec1=<$HUMAN>;
        my $humanrec2=<$HUMAN>;
        chomp $humanrec1;
        chomp $humanrec2;

        #get alignment scores
        my @AS_m1 = ($mouserec1 =~ m/AS:i:([0-9]*)/);
        my @AS_m2 = ($mouserec2 =~ m/AS:i:([0-9]*)/);
        my @AS_h1 = ($humanrec1 =~ m/AS:i:([0-9]*)/);
        my @AS_h2 = ($humanrec1 =~ m/AS:i:([0-9]*)/);

        my @hfields1=split /\t/,$humanrec1;
        my @hfields2=split /\t/,$humanrec2;

        # check that query names are the same
        if (!((($mfields1[0] eq $mfields2[0]) && ($hfields1[0] eq $hfields2[0])) && $mfields1[0] eq $hfields1[0])){
            die "record mismatch :\n$mouserec1\n$humanrec1\n$mouserec2\n$humanrec2";
        }
        
        ###### distinguish between reads that have not aligned, partially aligned, and fully aligned
        if (($AS_m1[0] > 104 || $AS_m2[0] > 104) && $AS_h1[0] < 104 && $AS_h2[0] < 104)
        {
            print $HOST "$hfields1[0]\n";
            $hostCount++
        }
        elsif ($AS_m1[0] < 104 && $AS_m2[0] < 104 && ($AS_h1[0] > 104 || $AS_h2[0] > 104))
        {
            print $GRAFT "$hfields1[0]\n";
            $graftCount++
        }
        elsif ($AS_m1[0] > 104 && $AS_m2[0] > 104 && $AS_h1[0] > 104 && $AS_h2[0] > 104)
        {
            print $BOTH "$hfields1[0]\n";
            $bothCount++
        }
        elsif ($AS_m1[0] < 10 && $AS_m2[0] < 10 && $AS_h1[0] < 10 && $AS_h2[0] < 10)
        {
            print $NEITHER "$hfields1[0]\n";
            $neitherCount++
        }
        else
        {
            print $AMBIGUOUS "$hfields1[0]\n";
            $ambiguousCount++
        }
        ###### print non-mouse only reads to sam file
        $totalCount++;
}

##### calculate class percentages
my $percentageH = ($hostCount / $totalCount) * 100;
my $percentageG = ($graftCount / $totalCount) * 100;
my $percentageB = ($bothCount / $totalCount) * 100;
my $percentageN = ($neitherCount / $totalCount) * 100;
my $percentageA = ($ambiguousCount / $totalCount) * 100;

$percentageH = sprintf("%.2f", $percentageH);
$percentageG = sprintf("%.2f", $percentageG);
$percentageB = sprintf("%.2f", $percentageB);
$percentageN = sprintf("%.2f", $percentageN);
$percentageA = sprintf("%.2f", $percentageA);

##### display results
print "Percentage of Mouse Reads: $percentageH\n";
print "Percentage of Human Reads: $percentageG\n";
print "Percentage of Both Reads: $percentageB\n";
print "Percentage of Neither Reads: $percentageN\n";
print "Percentage of Ambiguous Reads: $percentageA\n";
    
close $MOUSE;
close $HUMAN;
close $GRAFT;
close $HOST;
close $BOTH;
close $AMBIGUOUS;
close $NEITHER;
# close $OUT;

File: classifyByAS.py
# check number of arguments passed is correct
def checkArguments():
	if len(sys.argv) != 2:
		print "Usage: python classifyByAS.py mouse.bam human.bam ouput_dir"
		sys.exit()

def initializeVariables():
	# assign arguments to script variables
	mouse_bam=sys.argv[1]
	human_bam=sys.argv[2]
	# create output files
	graft = open("{:s}/graft.txt".format(sys.argv[3]),"w")
	host = open("{:s}/host.txt".format(sys.argv[3]),"w")
	both = open("{:s}/both.txt".format(sys.argv[3]),"w")
	ambiguous = open("{:s}/ambiguous.txt".format(sys.argv[3]),"w")
	neither = open("{:s}/neither.txt".format(sys.argv[3]),"w")
	# initialize counters
	total_count = 0; 
	host_count = 0; 
	graft_count = 0; 
	both_count = 0; 
	neither_count = 0;
	ambiguous_count = 0; 
	return (mouse_bam, human_bam, graft, host, both, ambiguous, neither, mouse_filtered, human_filtered, total_count,
		host_count, graft_count, both_count, neither_count, ambiguous_count)

def getReads(mouse_filtered, human_filtered, count):
	mouse_read_1 = mouse_filtered[count]
	mouse_read_2 = mouse_filtered[count+1]
	human_read_1 = human_filtered[count]
	human_read_2 = human_filtered[count+1]
	return mouse_read_1, mouse_read_2, human_read_1, human_read_2

def getAlignmentScores(mouse_read_1, mouse_read_2, human_read_1, human_read_2):
	AS_mouse_read_1 = re.findall('(?<=AS:i:)([0-9]*)', mouse_read_1[0])[0]
 	AS_mouse_read_2 = re.findall('(?<=AS:i:)([0-9]*)', mouse_read_2[0])[0]
 	AS_human_read_1 = re.findall('(?<=AS:i:)([0-9]*)', human_read_1[0])[0]
 	AS_human_read_2 = re.findall('(?<=AS:i:)([0-9]*)', human_read_2[0])[0]
 	return AS_mouse_read_1, AS_mouse_read_2, AS_human_read_1, AS_human_read_2

def getNames(mouse_read_1, mouse_read_2, human_read_1, human_read_2):
	mouse_read_1_name = mouse_read_1.split(' ')[0]
	mouse_read_2_name = mouse_read_2.split(' ')[0]
	human_read_1_name = human_read_1.split(' ')[0]
	human_read_2_name = human_read_2.split(' ')[0]

	return mouse_read_1_name, mouse_read_2_name, human_read_1_name, human_read_2_name

def checkReadNames(mouse_read_1_name, mouse_read_2_name, human_read_1_name, human_read_2_name):
	if mouse_read_1_name == mouse_read_2_name and human_read_1_name == human_read_2_name and mouse_read_1_name == human_read_1_name:
 		sys.exit('read names do not match: M1:{:s} M2:{:s} H1:{:s} H2:{:s}\n'.format(mouse_read_1_name, mouse_read_2_name, human_read_1_name, human_read_2_name))

def classify(AS_mouse_read_1, AS_mouse_read_2, AS_human_read_1, AS_human_read_2, host, host_count, graft, graft_count,
	both, both_count, neither, neither_count, ambiguous, ambiguous_count):
	if (AS_mouse_read_1 > 104 or AS_mouse_read_2 > 104) and AS_human_read_1 < 104 and AS_human_read_2 < 104:
		host.write('{:s}\n'.format(human_read_1_name))
		host_count += 1
	elif AS_mouse_read_1 < 104 and AS_mouse_read_2 < 104 and (AS_human_read_1 > 104 or AS_human_read_2 > 104):
		graft.write('{:s}\n'.format(human_read_1_name))
		graft_count += 1
	elif AS_mouse_read_1 > 104 and AS_mouse_read_2 > 104 and AS_human_read_1 > 104 and AS_human_read_2 > 104:
		both.write('{:s}\n'.format(human_read_1_name))
		both_count += 1
	elif AS_mouse_read_1 < 10 and AS_mouse_read_2 < 10 and AS_human_read_1 < 10 and AS_human_read_2 < 10:
		neither.write('{:s}\n'.format(human_read_1_name))
		neither_count += 1
	else:
		ambiguous.write('{:s}\n'.format(human_read_1_name))
		ambiguous_count += 1
	return graft_count, host_count, both_count, neither_count, ambiguous_count

# calculate class percentages
def calculatePercentages(graft_count, host_count, both_count, ambiguous_count, neither_count, total_count):
	percentages = ['','','','','']
	percentages[0] = float(graft_count)/total_count
	percentages[1] = float(host_count)/total_count
	percentages[2] = float(both_count)/total_count
	percentages[3] = float(ambiguous_count)/total_count
	percentages[4] = float(neither_count)/total_count
	return percentages

def displayOutput(percentages):
	sys.stdout.write("Percentage of Reads in Each Class\n\nGraft:{:s}\nHost:{:s}\nBoth:{:s}\nNeither:{:s}\nAmbiguous:{:s}"
		.format(percentages[0],percentages[1],percentages[2],percentages[3],percentages[4],))

# output results
if __name__ == '__main__':
	
	# import required modules
	import sys
	import argparse
	import re

	# argument parser
	# parser = argparse.ArgumentParser(description='Classify reads as host, graft, both, neither, or ambiguous.')
	# parser.add_argument('-m', '--mouse_bam', help='bam file for reads aligned to mouse', type=str, required=True)
	# parser.add_argument('-h', '--human_bam', help='bam file for reads aligned to human', type=str, required=True)
	# parser.add_argument('-o', '--output_dir', help='output directory for output bam and fastq files', type=str, required=True)
	# args = parser.parse_args(sys.argv)
	# mouse_bam = 

	checkArguments()
	(mouse_bam, human_bam, graft, host, both, ambiguous, neither, mouse_filtered, human_filtered, total_count, host_count, 
		graft_count, both_count, neither_count, ambiguous_count) = initializeVariables()
	
	for count in range(0,mouse_filtered,2):

		mouse_read_1, mouse_read_2, human_read_1, human_read_2 = getReads(mouse_filtered, human_filtered, count)
 		AS_mouse_read_1, AS_mouse_read_2, AS_human_read_1, AS_human_read_2 = getAlignmentScores(mouse_read_1, mouse_read_2, human_read_1, human_read_2)
 		mouse_read_1_name, mouse_read_2_name, human_read_1_name, human_read_2_name = getNames(mouse_read_1, mouse_read_2, human_read_1, human_read_2)
		checkReadNames(mouse_read_1_name, mouse_read_2_name, human_read_1_name, human_read_2_name) 		
        graft_count, host_count, both_count, neither_count, ambiguous_count = classify(AS_mouse_read_1, AS_mouse_read_2, AS_human_read_1, AS_human_read_2, 
        host, host_count, graft, graft_count, both, both_count, neither, neither_count, ambiguous, ambiguous_count)
        total_count += 1

	percentages = calculatePercentages(graft_count, host_count, both_count, ambiguous_count, neither_count, total_count)
	displayOutput(percentages)
		